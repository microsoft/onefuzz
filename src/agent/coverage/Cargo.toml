[package]
name = "coverage"
version = "0.1.0"
authors = ["fuzzing@microsoft.com"]
license = "MIT"
edition = "2018"

[features]
default = []
symbol-filter = []  # Remove after impl'd

[dependencies]
anyhow = "1.0"
bincode = "1.3"
cpp_demangle = "0.3"
debugger = { path = "../debugger" }
dunce = "1.0"
fixedbitset = "0.3"
goblin = "0.2"
iced-x86 = { version = "1.1", features = ["decoder", "op_code_info", "instr_info", "masm"] }
log = "0.4"
memmap2 = "0.2"
msvc-demangler = "0.8"
regex = "1.4"
rustc-demangle = "0.1"
serde = { version = "1.0", features = ["derive"] }
uuid = { version = "0.8", features = ["guid"] }
win-util = { path = "../win-util" }

[target.'cfg(target_os = "windows")'.dependencies]
pdb = "0.7"
winapi = "0.3"

[target.'cfg(target_os = "linux")'.dependencies]
pete = "0.4"
procfs = "0.9"

[dev-dependencies]
env_logger = "0.8"
serde_json = { version = "1.0", features = ["preserve_order"] }
structopt = "0.3"
