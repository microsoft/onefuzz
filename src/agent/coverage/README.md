# `coverage`

## Summary

The `coverage` crate is a library that provides components for recording binary code
coverage for userspace targets. The binary modules under test do not require static
instrumentation of any kind, but coverage will only be recorded for executable modules
that have debuginfo.

## Usage

### Example Tool

The `record` example demonstrates comprehensive usage of binary coverage recording and
conversion to source. It can be built from the `coverage` crate root via the command
`cargo build --examples --release.`

As an example, suppose you had a target name `app.exe`, with a directory of PNG test cases in `corpus`.

Binary coverage for a single specific input `example.png` could be recorded with the command:

```
record.exe -- ./app.exe corpus/example.png
```

The combined coverage for all inputs in the corpus can be recorded using the `--input-dir`/`-d` option:
```
record.exe -d corpus -- ./app.exe '@@'
```

In this case, the command after `--` is invoked multiple times. For each invocation, the
special `@@` input marker is replaced with the path to an input in `corpus`. The example
binary then merges the per-input coverage to produce the aggregated result.

To emit source + line coverage, just specify the `--output`/`-o` option:

```
record.exe -o source -d corpus -- ./app.exe '@@'
```

For Cobertura XML:

```
record.exe -o cobertura -d corpus -- ./app.exe '@@'
```

See `record.exe -h` for more options.

### Recording

The core type used for recording is `record::CoverageRecorder`. This accepts a Rust
standard library `Command`, and invokes it as a debuggee. Targets must exit before the
timeout, or no coverage will be returned. The output of recording is the `Recorded`
struct. This contains both an `output` field (target exit status and captured stdio
streams) and a `coverage` field. The `coverage` field contains the binary code coverage
organized by module and module-relative image offset.

### Allowlists

By default, coverage is recorded for all runtime-observed modules with debuginfo, and any
source file referred to by that debuginfo. Two allowlists can be used to control which
modules and source files have their coverage recorded.

Each allowlist is a flat text file with a simple syntax for path-matching rules.

- `/` or `\`-separated literal paths include an item (module or source file) exactly.
- `*` glob characters can be included anywhere, including within path components.
- Path patterns can be _excluded_ via the syntax `! <rule>`.
- Comments are supported using the `#` character.

If no allowlist is provided, the default allowlist contains only the rule `*`, which
includes all paths. If an allowlist is provided, then the default allow-all rule is
omitted. Files are then included only if they match an include rule but don't _also_ match
an (overriding) exclude rule.

An example source allowlist:

```
# 1. Record coverage for source files application root.
app/*

# 2. Also include library code, factored out of the application proper.
lib/*

# 3. But do _not_ record coverage for vendored library code.
! lib/vendor/*
```

With the above rules, we would have the following inclusion behavior:
- Include `src/main.c` (matches (1))
- Include `lib/utility.c` (matches (2))
- Exclude `lib/vendor/json.c` (matches (3), an exclude rule)
- Exclude `other/stuff.c` (does not match any allow rule)

### Source Coverage

#### Source File and Line

Source coverage is derived from binary coverage using debuginfo. The
`source::binary_to_source_coverage()` function converts a `BinaryCoverage` value to a
`SourceCoverage`, which describes the input binary coverage in terms of source files and
lines.

#### Cobertura XML

To obtain source and line coverage in the Cobertura XML format, you can directly convert a
`SourceCoverage` value to a `CoberturaCoverage` value using the `From` trait. The result
is serializable via the `CoberturaCoverage::to_string()` method. The conversion defined in
the `cobertura` module emits Cobertura designed to produce sensible HTML reports when
consumed by the ReportGenerator project.

## FAQ

1 _My Linux target uses dynamic linking or loading and has zero coverage info for shared libraries._

Make sure that your target shared libraries were compiled with debuginfo. If not, no
coverage will be measured for them at all. If coverage locations are defined, but never
reached, the shared libraries may not be found by the dynamic linker/loader at runtime.
This is an issue with your command invocation. You can debug this with the `record` tool's
`--dump-stdio` tool. A typical fix is to include the directories of non-system shared
libraries in the `LD_LIBRARY_PATH` environment variable.

2 _Coverage is being recorded generally, but the branches of `switch` statements don't seem reachable._

Please report your case to the OneFuzz team. Large `switch` statements are frequently
compiled to indirect jumps via tables, and we are working to improve coverage in these
cases.

3 _I have a source line with multiple statements. How do I know which ones are being hit or missed?_

Binary coverage is both ground truth and the most granular coverage format. The source
coverage representations do not currently include column info. If any module offset that
maps to a source line is missed, then the entire line is considered missed. In the future,
we intend to support partial line coverage.

4 _The source coverage reported for my target looks strange or incomplete._

Report your case to the OneFuzz team. Source coverage for optimized builds can be lossy,
and we are constrained by what info we get from the debuginfo. If possible, try recording
coverage for an unoptimized build of your target, and see if the same issue occurs. Either
way, we are always looking for edge cases we could use to drive improvements or
workarounds.
