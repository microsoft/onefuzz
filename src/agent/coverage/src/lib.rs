// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
#![allow(clippy::as_conversions)]
#![allow(clippy::new_without_default)]

#[cfg(target_os = "windows")]
mod intel;

#[cfg(target_os = "windows")]
pub mod pdb;

#[cfg(target_os = "windows")]
pub mod pe;

pub mod block;
pub mod cache;
pub mod code;
pub mod demangle;
pub mod report;
pub mod sancov;

#[cfg(target_os = "linux")]
pub mod disasm;

pub mod filter;
mod region;

#[cfg(test)]
mod test;
