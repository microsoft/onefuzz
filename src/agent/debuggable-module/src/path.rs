// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use std::ffi::OsStr;
use std::fmt;
use std::path::{Path, PathBuf};

use anyhow::{bail, Result};

/// Path to a file. Guaranteed UTF-8.
#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct FilePath(String);

impl FilePath {
    pub fn new(path: impl Into<String>) -> Result<Self> {
        let path = path.into();

        if Path::new(&path).file_name().is_none() {
            bail!("module path has no file name");
        }

        if Path::new(&path).file_stem().is_none() {
            bail!("module path has no file stem");
        }

        Ok(Self(path))
    }

    pub fn with_extension(&self, extension: impl AsRef<str>) -> Self {
        let path = self
            .as_path()
            .with_extension(extension.as_ref())
            .to_string_lossy()
            .into_owned();

        Self(path)
    }

    pub fn as_path(&self) -> &Path {
        Path::new(&self.0)
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }

    pub fn file_name(&self) -> &str {
        // Unwraps checked by ctor.
        Path::new(&self.0).file_name().unwrap().to_str().unwrap()
    }

    pub fn directory(&self) -> &str {
        // Unwraps checked by ctor.
        Path::new(&self.0).parent().unwrap().to_str().unwrap()
    }

    pub fn base_name(&self) -> &str {
        // Unwraps checked by ctor.
        Path::new(&self.0).file_stem().unwrap().to_str().unwrap()
    }
}

impl From<FilePath> for String {
    fn from(path: FilePath) -> Self {
        path.0
    }
}
impl From<FilePath> for PathBuf {
    fn from(path: FilePath) -> Self {
        path.0.into()
    }
}

impl AsRef<str> for FilePath {
    fn as_ref(&self) -> &str {
        self.as_str()
    }
}

impl AsRef<OsStr> for FilePath {
    fn as_ref(&self) -> &OsStr {
        self.as_str().as_ref()
    }
}

impl AsRef<Path> for FilePath {
    fn as_ref(&self) -> &Path {
        self.as_path()
    }
}

impl fmt::Display for FilePath {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}
