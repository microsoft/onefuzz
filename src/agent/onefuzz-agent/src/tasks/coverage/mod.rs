// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

#[cfg(any(target_os = "linux", target_os = "windows"))]
pub mod generic;
#[cfg(any(target_os = "linux", target_os = "windows"))]
pub mod libfuzzer_coverage;
#[cfg(any(target_os = "linux", target_os = "windows"))]
pub mod recorder;
pub mod total;
