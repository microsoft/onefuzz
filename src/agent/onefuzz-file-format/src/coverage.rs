// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

pub mod binary;
pub mod source;
