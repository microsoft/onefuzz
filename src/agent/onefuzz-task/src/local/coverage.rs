// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use std::{collections::HashMap, path::PathBuf};

use crate::{
    local::common::{
        build_local_context, get_cmd_arg, get_cmd_env, get_cmd_exe, get_synced_dir,
        get_synced_dirs, CmdType, CHECK_FUZZER_HELP, COVERAGE_DIR, INPUTS_DIR, READONLY_INPUTS,
        TARGET_ENV, TARGET_EXE, TARGET_OPTIONS, TARGET_TIMEOUT,
    },
    tasks::{
        config::CommonConfig,
        coverage::generic::{Config, CoverageTask},
    },
};
use anyhow::Result;
use async_trait::async_trait;
use clap::{Arg, ArgAction, Command};
use flume::Sender;
use onefuzz::syncdir::SyncedDir;
use schemars::JsonSchema;
use storage_queue::QueueClient;

use super::{
    common::{SyncCountDirMonitor, UiEvent},
    template::{RunContext, Template},
};

pub fn build_coverage_config(
    args: &clap::ArgMatches,
    local_job: bool,
    input_queue: Option<QueueClient>,
    common: CommonConfig,
    event_sender: Option<Sender<UiEvent>>,
) -> Result<Config> {
    let target_exe = get_cmd_exe(CmdType::Target, args)?.into();
    let target_env = get_cmd_env(CmdType::Target, args)?;
    let mut target_options = get_cmd_arg(CmdType::Target, args);
    let target_timeout = args.get_one::<u64>(TARGET_TIMEOUT).copied();

    let readonly_inputs = if local_job {
        vec![
            get_synced_dir(INPUTS_DIR, common.job_id, common.task_id, args)?
                .monitor_count(&event_sender)?,
        ]
    } else {
        get_synced_dirs(READONLY_INPUTS, common.job_id, common.task_id, args)?
            .into_iter()
            .map(|sd| sd.monitor_count(&event_sender))
            .collect::<Result<Vec<_>>>()?
    };

    let coverage = get_synced_dir(COVERAGE_DIR, common.job_id, common.task_id, args)?
        .monitor_count(&event_sender)?;

    if target_options.is_empty() {
        target_options.push("{input}".to_string());
    }

    let config = Config {
        target_exe,
        target_env,
        target_options,
        target_timeout,
        coverage_filter: None,
        module_allowlist: None,
        source_allowlist: None,
        input_queue,
        readonly_inputs,
        coverage,
        common,
    };

    Ok(config)
}

pub async fn run(args: &clap::ArgMatches, event_sender: Option<Sender<UiEvent>>) -> Result<()> {
    let context = build_local_context(args, true, event_sender.clone()).await?;
    let config = build_coverage_config(
        args,
        false,
        None,
        context.common_config.clone(),
        event_sender,
    )?;

    let mut task = CoverageTask::new(config);
    task.run().await
}

pub fn build_shared_args(local_job: bool) -> Vec<Arg> {
    let mut args = vec![
        Arg::new(TARGET_EXE).long(TARGET_EXE).required(true),
        Arg::new(TARGET_ENV).long(TARGET_ENV).num_args(0..),
        Arg::new(TARGET_OPTIONS)
            .long(TARGET_OPTIONS)
            .default_value("{input}")
            .value_delimiter(' ')
            .help("Use a quoted string with space separation to denote multiple arguments"),
        Arg::new(TARGET_TIMEOUT)
            .value_parser(value_parser!(u64))
            .long(TARGET_TIMEOUT),
        Arg::new(COVERAGE_DIR)
            .required(!local_job)
            .value_parser(value_parser!(PathBuf))
            .long(COVERAGE_DIR),
        Arg::new(CHECK_FUZZER_HELP)
            .action(ArgAction::SetTrue)
            .long(CHECK_FUZZER_HELP),
    ];
    if local_job {
        args.push(
            Arg::new(INPUTS_DIR)
                .long(INPUTS_DIR)
                .required(true)
                .value_parser(value_parser!(PathBuf)),
        )
    } else {
        args.push(
            Arg::new(READONLY_INPUTS)
                .required(true)
                .long(READONLY_INPUTS)
                .value_parser(value_parser!(PathBuf))
                .num_args(1..),
        )
    }
    args
}

pub fn args(name: &'static str) -> Command {
    Command::new(name)
        .about("execute a local-only coverage task")
        .args(&build_shared_args(false))
}

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
pub struct Coverage {
    target_exe: PathBuf,
    target_env: HashMap<String, String>,
    target_options: Vec<String>,
    target_timeout: Option<u64>,
    module_allowlist: Option<String>,
    source_allowlist: Option<String>,
    input_queue: Option<PathBuf>,
    readonly_inputs: Vec<PathBuf>,
    coverage: PathBuf,
}

#[async_trait]
impl Template for Coverage {
    async fn run(&self, context: &RunContext) -> Result<()> {
        let ri: Result<Vec<SyncedDir>> = self
            .readonly_inputs
            .iter()
            .enumerate()
            .map(|(index, input)| context.to_sync_dir(format!("readonly_inputs_{index}"), input))
            .collect();

        let input_q = if let Some(w) = &self.input_queue {
            Some(context.monitor_dir(w).await?)
        } else {
            None
        };

        let coverage_config = crate::tasks::coverage::generic::Config {
            target_exe: self.target_exe.clone(),
            target_env: self.target_env.clone(),
            target_options: self.target_options.clone(),
            target_timeout: None,
            readonly_inputs: ri?,
            input_queue: input_q,
            common: CommonConfig {
                task_id: uuid::Uuid::new_v4(),
                ..context.common.clone()
            },
            coverage_filter: None,
            coverage: context.to_monitored_sync_dir("coverage", self.coverage.clone())?,
            module_allowlist: self.module_allowlist.clone(),
            source_allowlist: self.source_allowlist.clone(),
        };

        context
            .spawn(async move {
                let mut coverage =
                    crate::tasks::coverage::generic::CoverageTask::new(coverage_config);
                coverage.run().await
            })
            .await;
        Ok(())
    }
}
