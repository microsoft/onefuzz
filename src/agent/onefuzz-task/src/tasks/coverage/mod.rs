// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

pub mod generic;
pub mod libfuzzer_coverage;
pub mod recorder;
pub mod total;
