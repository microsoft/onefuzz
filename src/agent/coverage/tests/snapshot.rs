// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

#[test]
#[cfg(all(target_os = "windows", feature = "slow-tests"))]
fn windows_snapshot_tests() {
    use coverage::{binary::DebugInfoCache, source::Line, AllowList};
    use debuggable_module::path::FilePath;
    use std::fmt::Write;
    use std::time::Duration;

    insta::glob!("windows", "*.cpp", |path| {
        let file_name = path.file_name().unwrap();

        // locate appropriate compiler
        let mut cl_exe =
            cc::windows_registry::find("x86_64-pc-windows-msvc", "cl.exe").expect("finding cl.exe");

        // path will have \\?\ prefix, cl.exe doesn't like it
        let input_path = dunce::canonicalize(path).unwrap();

        // directory to compile into
        let build_in = tempfile::tempdir().expect("creating tempdir");

        let output = cl_exe
            .arg("/EHsc")
            .arg("/Zi")
            .arg("/O2")
            .arg(&input_path)
            .current_dir(build_in.path())
            .spawn()
            .expect("launching compiler")
            .wait_with_output()
            .expect("waiting for compiler to finish");

        assert!(output.status.success(), "cl.exe failed: {:?}", output);

        let exe_name = {
            let mut cwd = build_in.path().to_path_buf();
            cwd.push(file_name);
            cwd.set_extension("exe");
            cwd
        };

        // filter to just the input test file:
        let source_allowlist = AllowList::parse(&input_path.to_string_lossy()).unwrap();

        let exe_cmd = std::process::Command::new(&exe_name);
        let recorded = coverage::CoverageRecorder::new(exe_cmd)
            .timeout(Duration::from_secs(120))
            .debuginfo_cache(DebugInfoCache::new(source_allowlist.clone()))
            .record()
            .unwrap();

        // generate source-line coverage info
        let source =
            coverage::source::binary_to_source_coverage(&recorded.coverage, &source_allowlist)
                .expect("binary_to_source_coverage");

        let file_coverage = source
            .files
            .get(&FilePath::new(input_path.to_string_lossy()).unwrap())
            .expect("coverage for input");

        let mut result = String::new();

        let file_source = std::fs::read_to_string(input_path).expect("reading source file");
        for (ix, content) in file_source.lines().enumerate() {
            let line = Line::new((ix + 1) as u32).unwrap();
            let prefix = if file_coverage.lines.contains_key(&line) {
                "[âœ”]"
            } else {
                "[ ]"
            };

            writeln!(result, "{prefix} {content}").unwrap();
        }

        insta::assert_snapshot!(result);
    });
}
