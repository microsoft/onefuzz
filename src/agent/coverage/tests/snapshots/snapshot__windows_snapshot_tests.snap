---
source: coverage/tests/snapshot.rs
expression: result
input_file: coverage/tests/windows/Inlinee.cpp
---
[ ] #include <iostream>
[ ] 
[ ] __declspec(dllexport) void test();
[ ] 
[ ] int main()
[✔] {
[✔]     std::cout << "Before\n";
[✔]     test();
[✔]     std::cout << "After\n";
[✔] }
[ ] 
[ ] __declspec(dllexport) void test() {
[✔]     std::cout << "Hello World!\n";
[ ] }

